proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -8893,52 +8818,98 @@ lemma active_sc_tcb_at_budget_sufficient:
   by (fastforce simp: budget_sufficient_def2 active_sc_tcb_at_def2 is_sc_active_kh_simp
                 dest: valid_refills_refill_sufficient active_implies_valid_refills)
 
+lemma awaken_body_valid_ready_qs:
+  "\<lbrace>\<lambda>s. valid_ready_qs s \<and> valid_release_q s \<and> active_sc_valid_refills s
+        \<and> release_queue s \<noteq> [] \<and> budget_ready (hd (release_queue s)) s\<rbrace>
+   awaken_body
+   \<lbrace>\<lambda>_. valid_ready_qs\<rbrace>"
+  apply (clarsimp simp: awaken_body_def tcb_release_dequeue_def bind_assoc)
+  apply (rule hoare_seq_ext[OF _ gets_sp], rename_tac rq)
+  apply (rule_tac B="\<lambda>_ s. valid_ready_qs s \<and> st_tcb_at runnable (hd rq) s
+                           \<and> released_sc_tcb_at (hd rq) s"
+               in hoare_seq_ext[rotated])
+   apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+    apply (wpsimp simp: valid_release_q_def obj_at_kh_kheap_simps)
+   apply (wpsimp simp: valid_release_q_def obj_at_kh_kheap_simps released_sc_tcb_at_def)
+   apply (fastforce intro!: active_sc_tcb_at_budget_sufficient
+                      simp: valid_release_q_def obj_at_kh_kheap_simps)
+  apply (wpsimp wp: possible_switch_to_valid_ready_qs
+              simp: released_sc_tcb_at_def)
+  done
+
+lemma awaken_body_valid_sched_action:
+  "\<lbrace>\<lambda>s. valid_sched_action s \<and> valid_release_q s \<and> active_sc_valid_refills s
+        \<and> release_queue s \<noteq> [] \<and> budget_ready (hd (release_queue s)) s\<rbrace>
+   awaken_body
+   \<lbrace>\<lambda>_. valid_sched_action\<rbrace>"
+  apply (clarsimp simp: awaken_body_def tcb_release_dequeue_def bind_assoc)
+  apply (rule hoare_seq_ext[OF _ gets_sp], rename_tac rq)
+  apply (rule_tac B="\<lambda>_ s. valid_sched_action s \<and> st_tcb_at runnable (hd rq) s
+                           \<and> released_sc_tcb_at (hd rq) s"
+               in hoare_seq_ext[rotated])
+   apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+    apply (wpsimp simp: valid_release_q_def obj_at_kh_kheap_simps)
+   apply wpsimp
+   apply (fastforce intro: active_sc_tcb_at_budget_sufficient
+                     simp: released_sc_tcb_at_def valid_release_q_def)
+  apply (wpsimp simp: obj_at_kh_kheap_simps)
+  done
+
+lemma fun_of_m_get_tcb_refill_budget_ready_valid_release_q:
+  "\<lbrakk>valid_release_q s; t \<in> set (release_queue s)\<rbrakk>
+   \<Longrightarrow> budget_ready t s = the (fun_of_m (get_tcb_refill_ready t) s)"
```

Comment:
It's hard to tell without actually going through the proofs, but my guess would be that this equality is more useful as
```
fun_of_m (get_tcb_refill_ready t) s = Some (budget_ready t s)